filesystem/flysystem/FlysystemAssetStore.php

```diff
@@ -180,58 +192,110 @@ protected function isSeekableStream($stream) {
+	protected function writeWithCallback($callback, $filename, $hash, $variant = null, $conflictResolution = null) {
+		// Set default conflict resolution
+		if(!$conflictResolution) {
+			$conflictResolution = $this->getDefaultConflictResolution($variant);
+		}
+		
+		// Validate parameters
+		if($variant && $conflictResolution === AssetStore::CONFLICT_RENAME) {
+			// As variants must follow predictable naming rules, they should not be dynamically renamed
+			throw new InvalidArgumentException("Rename cannot be used when writing variants");
+		}
+		if(!$filename) {
+			throw new InvalidArgumentException("Filename is missing");
+		}
+		if(!$hash) {
+			throw new InvalidArgumentException("File hash is missing");
+		}
+
 		$filename = $this->cleanFilename($filename);
-		$fileID = $this->getFileID($hash, $filename);
+		$fileID = $this->getFileID($filename, $hash, $variant);
 
 		// Check conflict resolution scheme
 		$resolvedID = $this->resolveConflicts($conflictResolution, $fileID);
-		if($resolvedID === false) {
-			// If defering to the existing file, return the sha of the existing file
-			$stream = $this
-				->getFilesystem()
-				->readStream($fileID);
-			$hash = $this->getStreamSHA1($stream);
-		} else {
+		if($resolvedID !== false) {
 			// Submit and validate result
 			$result = $callback($resolvedID);
 			if(!$result) {
 				throw new Exception("Could not save {$filename}");
 			}
-			
+
 			// in case conflict resolution renamed the file, return the renamed
 			$filename = $this->getOriginalFilename($resolvedID);
+			
+		} elseif(empty($variant)) {
+			// If defering to the existing file, return the sha of the existing file,
+			// unless we are writing a variant (which has the same hash value as its original file)
+			$stream = $this
+				->getFilesystem()
+				->readStream($fileID);
+			$hash = $this->getStreamSHA1($stream);
 		}
 
 		return array(
-			'Hash' => $hash,
 			'Filename' => $filename,
-			'Variant' => ''
+			'Hash' => $hash,
+			'Variant' => $variant
 		);
 	}
 
-	public function getMetadata($hash, $filename, $variant = null) {
-		$fileID = $this->getFileID($hash, $filename, $variant);
+	/**
+	 * Choose a default conflict resolution
+	 *
+	 * @param string $variant
+	 * @return string
+	 */
+	protected function getDefaultConflictResolution($variant) {
+		// If using new naming scheme (segment by hash) it's normally safe to overwrite files.
+		// Variants are also normally safe to overwrite, since lazy-generation is implemented at a higher level.
+		$legacy = $this->useLegacyFilenames();
+		if(!$legacy || $variant) {
+			return AssetStore::CONFLICT_OVERWRITE;
+		}
+		
+		// Legacy behaviour is to rename
+		return AssetStore::CONFLICT_RENAME;
+	}
+
+	/**
+	 * Determine if legacy filenames should be used. These do not have hash path parts.
+	 *
+	 * @return bool
+	 */
+	protected function useLegacyFilenames() {
+		return Config::inst()->get(__CLASS__, 'legacy_filenames');
```

Comment:
`get_class()` has the same issue though, right? `class MyFlysystemAssetStore extends FlysystemAssetStore` will return `MyFlysystemAssetStore`. And since the class is not `final`, and probably injectable, we can't use the instance class name.